### PR TITLE
feat: include uninstallation logic

### PR DESCRIPTION
**Pull request by alestiago**
## Status

**READY**

## Description

Adds the uninstallation logic.

Changes:
- New `uninstall` method in `CompletionInstallation`
- New `removeFrom` method in `ScriptConfigurationEntry`
- Defined `CompletionUnistallationException`

In a follow-up PR, I will add a new uninstall file to cache those uninstalled executables in order to avoid auto-installation installing them again.

In a follow-up PR, I will hook the new  `uninstall` method in `CompletionInstallation` as a CLI command the user can use.

In a follow-up PR, I will add logger information where needed for a better user experience.

## Type of Change

- [X] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: lib/src/installer/completion_installation.dart [hunk 1]
```diff
@@ -304,6 +304,82 @@ ${configuration!.sourceLineTemplate(scriptPath)}''';
...
     logger.info('Added config to $configFilePath');
   }
+
+  /// Uninstalls the completion for the command [executableName] on the current
+  /// shell.
+  ///
+  /// Before uninstalling, it checks if the completion is installed:
+  /// - The shell has an existing RCFile with a completion
+  /// [ScriptConfigurationEntry].
+  /// - The shell has an exisiting completion configuration file with a
```
Comment (renancaraujo) [inline lib/src/installer/completion_installation.dart:314]:
```suggestion
  /// - The shell has an exisiting completion configuration file with a
```

File: lib/src/installer/completion_installation.dart [hunk 2]
```diff
@@ -304,6 +304,82 @@ ${configuration!.sourceLineTemplate(scriptPath)}''';
...
+  /// [CompletionUnistallationException].
+  ///
+  /// Upon a successful uninstallation the executable [ScriptConfigurationEntry]
+  /// is removed from the shell configuration file. If after this removal the
+  /// latter is empty, it is deleted together with the the executable completion
+  /// script and the completion [ScriptConfigurationEntry] from the shell RC
+  /// file. In the case that there are no other completion scripts installed on
+  /// other shells the completion config directory is deleted, leaving the
+  /// user's system as it was before the installation.
+  void uninstall(String executableName) {
```
Comment (renancaraujo) [inline lib/src/installer/completion_installation.dart:327]:
```suggestion
  void uninstall(String rootCommand) {
```

File: lib/src/installer/exceptions.dart [hunk 3]
```diff
@@ -18,3 +18,24 @@ class CompletionInstallationException implements Exception {
...
+  CompletionUnistallationException({
+    required this.message,
+    required this.executableName,
+  });
+
+  /// The error message for this exception
+  final String message;
+
+  /// The command for which the installation failed.
+  final String executableName;
```
Comment (renancaraujo) [inline lib/src/installer/exceptions.dart:36]:
```suggestion
  final String rootCommand;
```

File: test/src/installer/completion_unistallation_exception_test.dart [hunk 4]
```diff
@@ -0,0 +1,60 @@
+import 'package:cli_completion/installer.dart';
+import 'package:test/test.dart';
+
+void main() {
```
Comment (renancaraujo) [inline test/src/installer/completion_unistallation_exception_test.dart:4]: file names should have the same structure are the tested files

File: lib/src/installer/exceptions.dart [hunk 5]
```diff
@@ -18,3 +18,24 @@ class CompletionInstallationException implements Exception {
   String toString() => 'Could not install completion scripts for $rootCommand: '
       '$message';
 }
+
+/// {@template completion_unistallation_exception}
+/// Describes an exception during the installation of completion scripts.
```
Comment (renancaraujo) [inline lib/src/installer/exceptions.dart:23]: Needs update

**Review by renancaraujo — APPROVED** [review]

File: lib/src/installer/completion_installation.dart (hunk 1 above)
Comment (alestiago) [reply lib/src/installer/completion_installation.dart:314]: What are you suggesting here?

File: lib/src/installer/script_configuration_entry.dart [hunk 6]
```diff
@@ -38,16 +38,46 @@ class ScriptConfigurationEntry {
...
+  /// Removes the entry with [name] from the [file].
+  ///
+  /// If the [file] does not exist, this will do nothing.
+  ///
+  /// If a file has multiple entries with the same [name], all of them will be
+  /// removed.
+  ///
+  /// If [shouldDelete] is true, the [file] will be deleted if it is empty after
+  /// removing the entry. Otherwise, the [file] will be left empty.
+  void removeFrom(File file, {bool shouldDelete = true}) {
```
[the remark this answers is not in the document]
Comment (alestiago) [reply lib/src/installer/script_configuration_entry.dart:64]: I'll change it to be false by default.

File: lib/src/installer/completion_installation.dart (hunk 1 above)
Comment (renancaraujo) [reply lib/src/installer/completion_installation.dart:314]: Github is crazy

**Review by renancaraujo — DISMISSED** [review]

**Pull request merged**